### ISSUE TITLE
[question] How to avoid shared library getting linked into the executable?

### ISSUE THREAD

**Issue by dragondive** [conan-io/conan#8309]
- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

I'm trying to create a demo to show use of Conan with a pre-built shared library. I made a small package using `export-pkg` which includes only the header file and the `.so` file as below:

    [random/0.1@aravind/testing]
    - include/
      +- random.h 
    - lib/
      +- librandom.so

I define a `conanfile.txt` as below to consume the package:

    [requires]
    random/0.1@aravind/testing

    [generators]
    cmake

    [options]
    random:shared=True

I use the `CMakeLists.txt` based on the [Getting Started](https://docs.conan.io/en/latest/getting_started.html) document:

    cmake_minimum_required(VERSION 3.18)
    project(random_use)

    add_definitions("-std=c++11")

    include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
    conan_basic_setup()

    add_executable(random_use main.cpp)
    target_link_libraries(random_use ${CONAN_LIBS})

Then I call:

    mkdir build && cd build
    conan install ..
    cmake ..
    cmake --build .

This links the shared library `librandom.so` into the executable. Instead I expected that it should be found at runtime (from `LD_LIBRARY_PATH`, for example as described in the document [How to manage shared libraries](https://docs.conan.io/en/latest/howtos/manage_shared_libraries.html). 

This does work as I expected on Windows. Without an `imports` to copy the `.dll` into the `bin` folder, the application crashes at runtime because the dll was not found. 

What am I missing here?

**Comment by dmn-star** [comment]
> This does work as I expected on Windows. Without an `imports` to copy the `.dll` into the `bin` folder, the application crashes at runtime because the dll was not found.
> 
> What am I missing here?

nothing so it works on windows

Standard Search Order for Desktop Applications

The standard DLL search order used by the system depends on whether safe DLL search mode is enabled or disabled. Safe DLL search mode places the user's current directory later in the search order.
Safe DLL search mode is enabled by default. To disable this feature, create the HKEY_LOCAL_MACHINE\System\CurrentControlSet\Control\Session Manager\SafeDllSearchMode registry value and set it to 0. Calling the SetDllDirectory function effectively disables SafeDllSearchMode while the specified directory is in the search path and changes the search order as described in this topic.
If SafeDllSearchMode is enabled, the search order is as follows:
**The directory from which the application loaded.**
The system directory. Use the GetSystemDirectory function to get the path of this directory.
The 16-bit system directory. There is no function that obtains the path of this directory, but it is searched.
The Windows directory. Use the GetWindowsDirectory function to get the path of this directory.
The current directory.
**The directories that are listed in the PATH environment variable.** Note that this does not include the per-application path specified by the App Paths registry key. The App Paths key is not used when computing the DLL search path.

https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order

**Comment by dmn-star** [comment]
>This links the shared library librandom.so into the executable. Instead I expected that it should be found at runtime (from LD_LIBRARY_PATH, for example as described in the document How to manage shared libraries.

Try to remove librandom.**a** as a quick workaround.

**Comment by memsharded** [comment]
``${CONAN_LIBS}`` contains all libraries that are listed in the dependencies. There could be different things you could do:

- In the recipe of the package, do not add the library to ``self.cpp_info.libs``. It will not be contained in the CONAN_LIBS var, and it will not be linked.
- Do not link with CONAN_LIBS var, but use specific CONAN_PKG::other_pkg variables, one per package you actually want to link with, but not ``random/0.1``.
- Use a more modern CMake generator, like ``cmake_find_package_multi``, instead of the ``cmake`` one. It will generate find.cmake scripts, one per dependency, and you do ``find_package()`` in your CMakeLists.txt, explicitly, for the dependencies you want to link.

Please let us know if this works, thanks!

**Comment by dragondive** [comment]
@memsharded Thanks for the hints. 

Turns out my previous understanding was wrong. After spending some time reading through the cmake documentation and examining the generated `conanbuildinfo.cmake`, the shared library *is* already loaded at runtime. I verified this by renaming the `librandom.so` file in the conan cache, then the application is unable to find it at runtime. If it was embedded into the executable as I had wrongly assumed, that wouldn't have happened. 

<s>In my simple example, the `random` package is the only one I'm linking. Removing the `target_link_libraries` from `CMakeLists.txt` causes linker error. It appears cmake wants to find the library at link time, and I cannot figure out how to tell cmake "don't link the library now, I will provide it at runtime".</s> The cmake documentation is not much helpful for a non-expert.

<s>As my (current) goal is to create a simple demo of Conan (and not cmake) for my colleagues, I will close this issue for now. I will try to look internally for a cmake expert later who can help me. </s>
